practice/flutter/flutter/logic.py

```diff
@@ -0,0 +1,167 @@
+    else:
+        new_flutt = models.Flutt(author=username, body=body, timestamp=time, author_id=user_id)
+        new_flutt.save()
+
+
+def get_current_time():
+    """Returns the server time at moment of access."""
+    return datetime.datetime.now()
+
+
+def get_all_flutts():
+    """Returns all posted Flutts.
+
+    >>> models.Flutt(author='Fred', body='2', timestamp=datetime.datetime(2016, 9, 2, 10, 28, 34), author_id=0).save()
+    >>> list(get_all_flutts())
+    [Flutt(author='Fred', body='2', timestamp=2016-09-02 10:28:34+00:00, author_id=0)]
+    """
+    return models.Flutt.objects.all()
+
+
+def sort_flutts_by_most_recent(flutts_list):
+    """Takes a list of Flutts, returns them sorted by most recent.
+
+    >>> sort_flutts_by_most_recent([models.Flutt(author='Fred', body='2', timestamp=datetime.datetime\
+    (2016, 9, 2, 10, 28, 34), author_id=0), models.Flutt(author='Bob', body='1', timestamp=datetime.datetime\
+    (2016, 9, 2, 10, 28, 35), author_id=1)])
+    [Flutt(author='Bob', body='1', timestamp=2016-09-02 10:28:35, author_id=1), Flutt(author='Fred', body='2', timestam\
+p=2016-09-02 10:28:34, author_id=0)]
+    """
+    return sorted(flutts_list, key=lambda Flutt: Flutt.timestamp, reverse=True)
+
+
+def get_matches_by_search_text(search_text):
+    """Takes a search text, searches through all posted Flutts for Flutts that contain that text, returns a list of
+    Flutts that match.
+
+    >>> models.Flutt(author='Fred', body='2', timestamp=datetime.datetime(2016, 9, 2, 10, 28, 34), author_id=0).save()
+    >>> get_matches_by_search_text(2)
+    <QuerySet [Flutt(author='Fred', body='2', timestamp=2016-09-02 10:28:34+00:00, author_id=0)]>
+    >>> get_matches_by_search_text(3)
+    Traceback (most recent call last):
+    ...
+    LookupError: No matches
+    """
+    matches = models.Flutt.objects.filter(body__contains=search_text)
+    if not matches:
+        raise LookupError('No matches')
+    else:
+        return matches
+
+
+def login_user(request, username, password):
+    """Logs the user in.
+
+    Returns a LookupError if the username does not exist.
+
+    Returns a ValueError if the password was incorrect.
+
+    >>> user = User.objects.create_user('john', 'lennon@thebeatles.com', 'johnpassword')
+    >>> c = client.Client()
+    >>> c.login(username='john', password='johnpassword')
+    True
+    >>> login_user('request', 'john', 'floorb')
+    Traceback (most recent call last):
+    ...
+    ValueError: Incorrect password
+    >>> login_user('request', 'fee', 'foo')
+    Traceback (most recent call last):
+    ...
+    LookupError: That user does not exist.
+    """
+    user = authenticate(username=username, password=password)
+    if user is not None:
+        login(request, user)
+    try:
+        User.objects.get(username=username)
+    except User.DoesNotExist:
+        raise LookupError('That user does not exist.')
+    else:
+        raise ValueError('Incorrect password')
```

Comment:
This logic should live in views. The core business logic of flutter doesn't need to know how specifically to do authentication. That's part of the HTTP interface.

Minor point: many services, for security, don't reveal whether the username even exists, just that the combo was wrong. This makes it much harder to guess passwords.